### PR TITLE
Create optional provider config setting for using or disabling the child token

### LINKED ISSUE #29: Vault provider is unable to access data with a non-root token
_This issue was originally opened by @mperriere as hashicorp/terraform#16457. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Hi there,

### Terraform Version
0.10.6

### Terraform Configuration Files
```
cat main.tf
provider "vault" {
  address = "http://${var.vault_host}:${var.vault_port}"
  skip_tls_verify = true
  token = "dc985ea7-57eb-77b5-17c9-ae86fe019c82"
}

data "vault_generic_secret" "key" {
  path = "secret/dev/onenext/db_user01"
}

output pwd {
 value = "${data.vault_generic_secret.key.data["pwd"]}"
}

cat variable.tf
variable vault_host {
 #default="consuldev.pipeline.aws"
 default="10.196.14.160"
}
variable vault_port {
 default="8200"
}
variable env {
 default="dev"
}
variable project_name {
 default="onenext"
}

```

### Debug Output
terraform-0.10.6 apply
data.external.read_token: Refreshing state...
data.external.fetch_token: Refreshing state...
Error refreshing state: 1 error(s) occurred:

* provider.vault: failed to create limited child token: Error making API request.

URL: POST http://10.196.14.160:8200/v1/auth/token/create
Code: 403. Errors:

* permission denied

### Expected Behavior
I'm trying to create a policy based for every project and env: we need a token for every project, in every environnment (DEV, PREP, PROD).
Then i attach a token to it, write a secret inside the path defined in the policy, and finally i'm trying to read it from Terraform. The token should allow read access to the credentials, but it does not.

the rules are defined in a ansible playbook (outside terraform scope)
      path "secret/{{env}}/{{project_name}}/*" {
        capabilities = ["create", "read", "update", "delete", "list"]
      }
      path "secret/{{env}}/{{project_name}}" {
        capabilities = ["list"]
      }
      path "auth/token/lookup-self" {
        capabilities = ["read"]
      }

Next step, i'm creating a token attached to that policy:
(sorry, ansible code too):
- hashivault_token_create:
    display_name: "{{project_name}}-{{env}}"
    policies: ["{{project_name}}-{{env}}"]
    renewable: True
    token: "{{root_token}}"

This token is available, and can be queried on the Vault server:
VAULT_TOKEN=dc985ea7-57eb-77b5-17c9-ae86fe019c82 vault token-lookup --address=http://127.0.0.1:8200
Key                     Value
---                     -----
accessor                ec4ccd86-c6b9-b0e5-b337-5504ced231cd
creation_time           1508927130
creation_ttl            2764800
display_name            token-onenext-dev
explicit_max_ttl        0
id                      dc985ea7-57eb-77b5-17c9-ae86fe019c82
meta                    <nil>
num_uses                0
orphan                  false
path                    auth/token/create
policies                [default onenext-dev]
renewable               true
ttl                     2761029

Next step. i'm writing a secret in the policy path. No issue there.

Now i'm trying to read it from the vault cli: this is ok, i can read the secret. This proves that the secret, policy and token are correctly defined:

 VAULT_TOKEN=dc985ea7-57eb-77b5-17c9-ae86fe019c82 vault read --address=http://127.0.0.1:8200/ secret/dev/onenext/db_user01
Key                     Value
---                     -----
refresh_interval        768h0m0s
pwd                     tai.s0dohM5r

Now i'm trying to do the same thing with terraform.
I get that error:
* provider.vault: failed to create limited child token: Error making API request.
URL: POST http://10.196.14.160:8200/v1/auth/token/create
Code: 403. Errors:

PS: if i switch back to the root token in the provider vault definition, i can access to the secret.

It  seems that terraform is trying to create a temporary token instead of using the one newly created to just read the secret key with the provided token.

Perhaps i did not understand the documentation? I did not find full usage examples on the internet, i don't know what's wrong here.

### Steps to Reproduce
Please list the full steps required to reproduce the issue, for example:
2. `terraform apply`

### LINKED ISSUE #722: Feature Request: add provider argument `use_child_token` to control child token creation
### Terraform Version
v0.12.24

### Affected Resource(s)
- `vault_token`

### Expected Behavior
User can choose whether to use child token or use the injected token itself.

### Actual Behavior
One cannot use the injected token itself.
Since the usage of child token is default and cannot be refused, `vault_token` resources created with terraform will be revoked in 20minutes(default of `max_lease_ttl_seconds`.)
There are [periodic service tokens](https://learn.hashicorp.com/vault/security/tokens#periodic-service-tokens) which are meant to have possibility to live no max TTL.

For the developers using terraform to create vault resources, they might not be aware of the child token because the create `vault_token` resources won't show that their parent has a short TTL. I've been struggling while dealing with periodic tokens disappearing in 20 minutes without a notion.

### Steps to Create Periodic Tokens
```

resource "vault_token_auth_backend_role" "test_role" {
  role_name           = "test_role"
  token_period        = "2592000"
}

resource "vault_token" "periodic_token" {
  role_name = vault_token_auth_backend_role.test_role.role_name
}

```

### References
- https://learn.hashicorp.com/vault/security/tokens#periodic-service-tokens
- https://github.com/terraform-providers/terraform-provider-vault/issues/550#issuecomment-552360100

### PR DESCRIPTION
**Pull request by tstraley**
This adds a new (often requested) vault provider configuration setting which can allow users to disable the creation of the intermediate child token (strongly discouraged, but made available for the cases that need it).

```
provider "vault" {
    address = "https://vault.testdomain"
    create_intermediate_child_token = false
}
```

(Open to changing the name of this config item, or related env var, but I like to be verbose with settings that typically shouldn't be adjusted without reading the documentation).

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Closes #29 
Closes #722 
Relates #550 
Relates #368 
Relates #192 

Relates to the following as well:
https://github.com/hashicorp/terraform/issues/16457
https://github.com/hashicorp/terraform/pull/14839
https://groups.google.com/forum/#!topic/terraform-tool/wtlLrKVQlAo

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-vault/blob/master/CHANGELOG.md):

```release-note
Add `create_intermediate_child_token` optional provider configuration setting to control the use of ephemeral child tokens for use by terraform.
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run TestAccChildToken'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run TestAccChildToken -timeout 120m
?   	github.com/terraform-providers/terraform-provider-vault	[no test files]
?   	github.com/terraform-providers/terraform-provider-vault/cmd/coverage	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-vault/util	0.012s [no tests to run]
=== RUN   TestAccChildToken
--- PASS: TestAccChildToken (1.15s)
PASS
ok  	github.com/terraform-providers/terraform-provider-vault/vault	1.168s

```

### CODE REVIEW

**Comment by nikolaishields** [comment]
@tstraley Thankyou so much for the contribution! looks like the only thing holding this back is for you to sign the CLA as mentioned above by @hashicorp-cla. Don't want to be a bother, but me and my team could really use this once its merged upstream.

**Comment by nikolaishields** [comment]
After further investigation it appears that @tstraley has been largely inactive on github for 6+ months, with the only blocker being that they have not accepted the hashicorp CLA. I will be reopening this pull request with no change in content other than accepting the CLA. @tstraley Thank you for your contributions, I in no way intend to take credit for the work you've performed, and only want this feature for the sake of the community.

**Comment by tstraley** [comment]
@nikolaishields - thanks for the bumps, this had fallen way off my radar. I am now seeking approval from my employer's license review board before I can accept the hashicorp CLA and get this moved on.

**Comment by tstraley** [comment]
Can I please get reviews from the maintainers of this project? I believe this is ready for merge (long overdue). Thanks :)

**Comment by tstraley** [comment]
@catsby @tomhjp @tvoran @jasonodonnell - is this something any of you could look at?

**Comment by matttrach** [comment]
I think this could resolve several other feature requests, what do you think @tyrannosaurus-becks ?

**Comment by matttrach** [comment]
As for the security concern, in many use cases an encrypted Terraform backend could mitigate a lot of the risk.

**Comment by tyrannosaurus-becks** [comment]
Hi @matttrach , I'm no longer involved with this project and have moved on from HashiCorp.

**Comment by matttrach** [comment]
We will miss you @tyrannosaurus-becks !
@jasonodonnell would you mind taking a look?

**Comment by surendarkumavat** [comment]
@tstraley Many Thanks for your contribution on this. With the remote disk encrypted backend becoming the norm, this has become a much awaited feature. @maintainers any updates on the ETA?

**Comment by rmgpinto** [comment]
Can this me merged and released please?

**Comment by rmgpinto** [comment]
Or is there a way to merge this change locally and replace the hashicorp provider?

**Comment by rmgpinto** [comment]
@jasonodonnell can this be approved please? It's just a small change on 3 files.

**Comment by jasonodonnell** [comment]
Hi @rmgpinto, we're still discussing the security implications of this change internally and haven't reached a decision yet. This PR is high quality and it looks good from a code perspective, but the exposure of Vault tokens via Terraform state files and logs are definitely concerning (even when the defaults are security focused).

**Comment by rmgpinto** [comment]
Thanks for the quick feedback!

**Comment by alan2wong** [comment]
@jasonodonnell, could you provide some insight as to what the security implications are for this particular PR?

**Comment by tstraley** [comment]
@jasonodonnell giving this a bump.

I want to call out that all of the security considerations are well documented and this isn't something that users would use unless they need it and understand those risks. Having the option available has been a major request for many years now, and I'm not hearing any reason this should be held up further.

Please let me know if you'd like me to resolve the new merge conflicts and re-push so this can finally land.

**Comment by janmaghuyop** [comment]
@tstraley thank you for this. Just commenting here for an update, so I could just `terraform apply --auto-approve` all the way when approved.

**Comment by benashz** [comment]
@tstraley  we are taking another look at this PR. Would you mind rebasing it off of main?

Thanks,

Ben

**Comment by tstraley** [comment]
@benashz - just rebased, resolved conflicts, and pushed updated version up. Please have a look and let me know if you'd like any changes.

Happy :turkey:

File: vault/provider_test.go
Comment (benashz) [inline vault/provider_test.go:770]: I think this logic could probably be moved to the test, which would then just call `checkSelfToken()`.

File: website/docs/index.html.markdown
Comment (benashz) [inline website/docs/index.html.markdown:142]: For extra clarity it may be a good idea to update the `token` documentation which discusses "child" token creation.

File: website/docs/index.html.markdown
Comment (benashz) [inline website/docs/index.html.markdown:145]: This is only in the case when `TF_LOG` is set to `debug`. We should probably mention that here.

File: vault/provider_test.go
Comment (benashz) [inline vault/provider_test.go:708]: It looks like `CheckDestroy` may be being misused here. According to the [docs](https://github.com/hashicorp/terraform-plugin-sdk/blob/76b3563a7afe40aa29d08d1661c5502b59e3811a/helper/resource/testing.go#L341-L342) it is used to ensure that the all the provisioned resources have been properly removed.

In addition `CheckDestroy` is called conditionally
[here](https://github.com/hashicorp/terraform-plugin-sdk/blob/76b3563a7afe40aa29d08d1661c5502b59e3811a/helper/resource/testing_new.go#L67-L72), so there is no guarantee that the environment will be properly cleaned. up. This could pollute the environment of other tests. I suggest adding
```
defer testEnvCleanup("TERRAFORM_VAULT_CREATE_CHILD_TOKEN")
```
 to the test function's scope.

File: vault/provider_test.go
Comment (benashz) [inline vault/provider_test.go:632]: I think that we probably only need the `ExpectChildToken` and `TestChildTokenEnv` fields, all of the others can be dropped.

You can use `strconv.FormatBool(test.ExpectChildToken)` for use in the string context.

See:  https://pkg.go.dev/strconv#FormatBool for more details

**Review by benashz — CHANGES_REQUESTED** [review]
This is looking pretty good so far. I have provided some initial feedback.

Thanks,

Ben

File: vault/provider.go
Comment (tstraley) [reply vault/provider.go]: moved and renamed `skip_child_token`

File: vault/provider.go
Comment (tstraley) [reply vault/provider.go]: renamed to `VAULT_SKIP_CHILD_TOKEN`

File: vault/provider.go
Comment (tstraley) [reply vault/provider.go]: works for me. Done and done.

File: vault/provider_test.go
Comment (tstraley) [reply vault/provider_test.go]: > Please update the documentation to match.

Can you be more clear about what documentation needs updating here? The comment above this function is still 100% accurate, and the change to make this more generic didn't affect anything else.

Comment (tstraley) [reply vault/provider_test.go]: as for your second comment:

There are several tests that call this function. One of those tests always includes a header provider config, the others will only want the provider configuration if the test is specifically testing the provider schema.

To be fair, I think I agree that simply passing in an empty provider config string would be more logical, and possibly nicer within this function, but it would require moving that specific logic (to determine if the provider config should be an empty string or actually contain the schema under test) elsewhere, and then we're back to this same conditional. If you would prefer I separate this into two functions, one for generating the Terraform config string and one for evaluating a conditional to decide if the provider portion of that Terraform config should including something or not, let me know, but I've left this as-is for now, since it doesn't really make sense to me.

In case the confusion you had was around the naming of this function (eg. why does `testProviderConfig` sometimes not include the "Provider" portion of the config and only a data resource?) I changed the name to `getTerraformConfig` which reduces that confusion at the expense of being a bit generic.

File: vault/provider_test.go
Comment (tstraley) [reply vault/provider_test.go]: I'm not sure I agree with your reasoning (functions should be able to describe what they do with non-opaque names), but I changed to `checkSelfToken` as that seems reasonable to me.

File: vault/provider_test.go
Comment (tstraley) [reply vault/provider_test.go:765]: I'm not sure what you are asking for here. Can you elaborate, please?

In one interpretation I take your meaning to be that this logic can live in `CheckSelfToken` directly, but because that function is used by other tests that need to check their own specific attributes with their own expectations, this conditional based on whether we are using the child token or not doesn't apply to those tests. In another interpretation I think you may be asking for this to be an anonymous function within either the test case struct, or the `Check` of the test step, but I would argue using an anonymous function in that setting is less clear and readable.

Let me know if I'm missing your meaning.

File: website/docs/index.html.markdown
Comment (tstraley) [reply website/docs/index.html.markdown]: updated :+1:

File: website/docs/index.html.markdown
Comment (tstraley) [reply website/docs/index.html.markdown]: Done.

File: vault/provider_test.go
Comment (tstraley) [reply vault/provider_test.go]: Thanks, that's a fair call. Moved this into the test functions as a direct env var unset.

File: vault/provider_test.go
Comment (tstraley) [reply vault/provider_test.go:631]: I don't think so.
There are two configuration methods being tested in the parameterization: the schema (provider config block in .tf file) and/or the environment variables that manage the same configuration. Because we want to be able to test them independently as well as together, we need a way to state whether we want to **use** the schema, the env var, both, or neither.

I have updated the names of these fields to be more clear that they set whether to **use** a particular config method, and then assign the `skip_child_token` setting separately. Please note that each of these fields is independent and not always the same bool as any other field.

File: vault/provider_test.go
Comment (benashz) [reply vault/provider_test.go]: The extra boolean is argument is redundant with the config string.

`if config != "" {...}`

File: vault/provider_test.go
Comment (benashz) [reply vault/provider_test.go:765]: Rather than creating a new function call `checkSelfToken()` directly with the source attribute and the expected value.

Comment (tstraley) [reply vault/provider_test.go:765]: Could you provide some example code to illustrate how you expect this to look and work please? I must be missing something. Right now the attribute and expected value are based on whether we expect a child token should have been created or not; thus this if-else case. Since we do not know how to call `checkSelfToken` without switching on this condition, this logic needs to occur somewhere, rather than calling "directly" to the `checkSelfToken` function.

File: vault/provider_test.go
Comment (benashz) [inline vault/provider_test.go:724]:
```suggestion
func testProviderConfig(includeProviderConfig bool, config string) string {
```

**Review by benashz — CHANGES_REQUESTED** [review]
Looking good! Just a few minor suggestions.

File: vault/provider_test.go
Comment (tstraley) [reply vault/provider_test.go]: Take a look at line 621: https://github.com/hashicorp/terraform-provider-vault/pull/775/files#diff-96e3c1e3f2b69ac273388431a9cebb277d7a5d596691beb0a32b5821c5d6b063R621
and 703: https://github.com/hashicorp/terraform-provider-vault/pull/775/files#diff-96e3c1e3f2b69ac273388431a9cebb277d7a5d596691beb0a32b5821c5d6b063R703

In both of those calls, config is never `""`. So this is NOT redundant.

I can refactor this (and the existing tests that use it) even further if you only want this function to have one argument, but to keep it clean and readable, I will probably need to add 2 additional functions then call this `getTerraformConfig` in order to keep each test's logic isolated. Is that what you are asking for?

File: vault/provider.go
Comment (tstraley) [reply vault/provider.go:77]: The other vars don't have `TERRAFORM_` prefix either. Should this really have it?

File: vault/provider_test.go
Comment (tstraley) [reply vault/provider_test.go:631]: This simply followed a pattern already set by other tests in here like the existing `TestAccTokenName`. I'm certainly not going to remove valid test cases at this point - more coverage is better.

File: vault/provider.go
Comment (benashz) [reply vault/provider.go:77]: Variable names with the `VAULT_` prefix are owned by Vault itself e.g. `VAULT_ADDR`, `VAULT_TOKEN`.  The variable in question is provider specific, and we wouldn't want it to potentially affect the Vault configuration in any way.

File: vault/provider_test.go
Comment (benashz) [reply vault/provider_test.go:765]: I was suggesting that we move `checkUsedToken` to be a closure within the related test and renaming it to `checkTokenUsed`:

```
@@ -628,6 +629,19 @@ func TestAccTokenName(t *testing.T) {
 
 func TestAccChildToken(t *testing.T) {
        defer os.Unsetenv("VAULT_SKIP_CHILD_TOKEN")
+
+       checkTokenUsed := func(expectChildToken bool) resource.TestCheckFunc {
+               if expectChildToken {
+                       // If the default child token was created, we expect the token
+                       // used by the provider was named the default "token-terraform"
+                       return checkSelfToken("display_name", "token-terraform")
+               } else {
+                       // If the child token setting was disabled, the used token
+                       // should match the user-provided VAULT_TOKEN
+                       return checkSelfToken("id", os.Getenv("VAULT_TOKEN"))
+               }
+       }
+
        tests := []struct {
                skipChildTokenEnv    string
                useChildTokenEnv     bool
@@ -701,7 +715,7 @@ func TestAccChildToken(t *testing.T) {
                                                }
                                        },
                                        Config: getTerraformConfig(test.useChildTokenSchema, `skip_child_token = `+test.skipChildTokenSchema),
-                                       Check:  checkUsedToken(test.expectChildToken),
+                                       Check:  checkTokenUsed(test.expectChildToken),
                                },
                        },
                })
@@ -762,18 +776,6 @@ func checkSelfToken(attrName string, expectedValue string) resource.TestCheckFun
        }
 }
 
-func checkUsedToken(expectChildToken bool) resource.TestCheckFunc {
-       if expectChildToken {
-               // If the default child token was created, we expect the token
-               // used by the provider was named the default "token-terraform"
-               return checkSelfToken("display_name", "token-terraform")
-       } else {
-               // If the child token setting was disabled, the used token
-               // should match the user-provided VAULT_TOKEN
-               return checkSelfToken("id", os.Getenv("VAULT_TOKEN"))
-       }
-}
-
```

File: vault/provider_test.go
Comment (benashz) [reply vault/provider_test.go:631]: Agreed, it's fine to keep it as-is for now.

File: vault/provider_test.go
Comment (benashz) [reply vault/provider_test.go]: Thanks, but I think it is okay to keep things as they are, but `getTerraformConfig` should be renamed as suggested earlier.

Comment (benashz) [reply vault/provider_test.go]: Thanks, but I think it is okay to keep things as they are, but `getTerraformConfig` should be renamed as suggested earlier.

File: vault/provider_test.go
Comment (tstraley) [reply vault/provider_test.go]: I'm happy with that. Done.

File: vault/provider.go
Comment (tstraley) [reply vault/provider.go]: Got it, thanks for the explanation. Fixed.

**Review by benashz — CHANGES_REQUESTED** [review]

**Review by benashz — APPROVED** [review]
So we just need to address the documentation comment, then +1.

Thank you for your contribution to HashiCorp!

**Comment by benashz** [comment]
Thanks again @tstraley for all the effort on this PR. It's much appreciated!

**Pull request merged**
